### ISSUE TITLE
TypeError when Livepush rebuilds one service of several

### ISSUE THREAD

**Issue by kb2ma** [balena-io/balena-cli#2724]
# Expected Behavior
Livepush should not generate an error when modifying a Dockerfile in the attached setup.

[type-error.zip](https://github.com/balena-io/balena-cli/files/14070736/type-error.zip)

# Actual Behavior
When modify one of the Dockerfiles, see a message like below.

```
$ balena push 192.168.1.1
...
[Live]    Device state settled
```
Modify Dockerfile
```
[Live]    Detected Dockerfile change, performing full rebuild of service alpine
...
[Build]   [alpine] Successfully built 33d946754ae7
[Build]   [alpine] Successfully tagged local_alpine:latest
[Logs]    [2024-01-27T02:07:40.884Z] Service exited 'alpine sha256:387ac3ddc40b1254e4107bdf7ea520df7673de4767a8a0ce1b6e6059e8263c7e'
[Error]   There was an error rebuilding the service: TypeError: Cannot read properties of undefined (reading 'external')
```
Collected a stack trace for the error:
```
[Debug]   TypeError: Cannot read properties of undefined (reading 'external')
[Debug]       at getImageNameFromTask (/home/kbee/dev/balena-cli/repo/build/utils/device/deploy.js:349:18)
[Debug]       at /home/kbee/dev/balena-cli/repo/build/utils/device/deploy.js:380:24
[Debug]       at /home/kbee/dev/balena-cli/repo/node_modules/lodash/lodash.js:4967:15
[Debug]       at baseForOwn (/home/kbee/dev/balena-cli/repo/node_modules/lodash/lodash.js:3032:24)
[Debug]       at /home/kbee/dev/balena-cli/repo/node_modules/lodash/lodash.js:4936:18
[Debug]       at Function.forEach (/home/kbee/dev/balena-cli/repo/node_modules/lodash/lodash.js:9410:14)
[Debug]       at generateTargetState (/home/kbee/dev/balena-cli/repo/build/utils/device/deploy.js:357:7)
[Debug]       at LivepushManager.handleServiceRebuild (/home/kbee/dev/balena-cli/repo/build/utils/device/live.js:231:76)
[Debug]       at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
[Debug]       at async LivepushManager.handleFSEvents (/home/kbee/dev/balena-cli/repo/build/utils/device/live.js:170:13)
```

# Analysis
The problem is that the service rebuild routine `handleServiceRebuild()` passes only the single task being rebuilt on this [line](https://github.com/balena-io/balena-cli/blob/4b2602676b098d32de4b16806a21cb45af0752f1/lib/utils/device/live.ts#L426). However, `generateTargetState()` expects all of the tasks in the passed composition on this [line](https://github.com/balena-io/balena-cli/blob/4b2602676b098d32de4b16806a21cb45af0752f1/lib/utils/device/deploy.ts#L641). So `undefined` is passed to `getImageNameFromTask()`, which generates the error.

The solution then may be to pass all tasks to `generateTargetState().`

# Specifications
- **balena CLI version:** 17.5.0
- **Operating system version:** Ubuntu 22.04

**Comment by kb2ma** [comment]
@pipex Do you see a straightforward fix here?

**Comment by pipex** [comment]
So there is a check right above for the task to set the contract https://github.com/balena-io/balena-cli/blob/4b2602676b098d32de4b16806a21cb45af0752f1/lib/utils/device/deploy.ts#L630-L635 but then task is obtained directly on the [line you linked](https://github.com/balena-io/balena-cli/blob/4b2602676b098d32de4b16806a21cb45af0752f1/lib/utils/device/deploy.ts#L641). 

That seems like a bug introduced on this PR https://github.com/balena-io/balena-cli/pull/

I don't think we can pass all the tasks to the `generateTargetState` function, because the `handleServiceRebuild` is probably being called by livepush for a specific service.

In order to fix the bug and keep the functionality of the PR i would try to find where the `composition` property is being created and add the image reference there, instead of adding it at the last minute. There might be a reason the PR author didn't go that route, but it probably will require some testing

**Comment by pipex** [comment]
> I don't think we can pass all the tasks to the generateTargetState function, because the handleServiceRebuild is probably being called by livepush for a specific service.

Actually, you are right  @kb2ma, the solution is probably passing all the build tasks to the `generateTargetState` function to ensure the image name is obtained from the right place

**Comment by pipex** [comment]
I still wonder why we are only hearing about this issue now, given that the PR is 3 years old

**Comment by kb2ma** [comment]
I found a [report](https://forums.balena.io/t/balena-livepush-cannot-rebuild-service-typeerror-cannot-read-property-external-of-undefined/350752) in the forums from a couple of years ago.

**Comment by kb2ma** [comment]
Fixed with v17.5.1.
